### PR TITLE
Add OAuth2 sample authenticator

### PR DESCRIPTION
**Pull request by Avarjana**
## Description
Add sample code for OAuth2GenericAuthenticator with Kakao. 

## Related issues: 
- https://github.com/wso2/product-is/issues/14166

## Related PRs: 
- https://github.com/wso2-extensions/identity-outbound-auth-oauth2/pull/16
- https://github.com/wso2/product-is/pull/14234

### CODE REVIEW

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/pom.xml [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+                        <Export-Package>
+                            !org.wso2.carbon.identity.sample.oauth2.federated.authenticator.internal,
+                            org.wso2.carbon.identity.sample.oauth2.federated.authenticator.*; version=${project.version}
+                        </Export-Package>
+                        <Import-Package>
+                            com.nimbusds.jose.util;version="${nimbusds.osgi.version.range}",
+                            net.minidev.json;version="${json-smart.version.range}"
+                            org.wso2.carbon.identity.application.authentication.framework.*;version="${carbon.identity.framework.version.range}",
+                        </Import-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
```
Comment (piraveena) [inline authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/pom.xml:114]: remove dynamic import

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    @Override
+    protected String getUserInfoEndpoint(Map<String, String> authenticatorProperties) {
+
+        return OAuth2CustomAuthenticatorConstants.INFO_URL;
+    }
+
+    // TODO: Override buildClaims to match the response of respective OAuth2 identity provider
+
+    @Override
+    public List<Property> getConfigurationProperties() {
```
Comment (piraveena) [inline authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java:69]: Don't we need userinfo URL

Comment (piraveena) [reply authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java:69]: we can either hard code them or get them from the config. Let's add a comment if they want to get them from UI config

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    @Activate
+    protected void activate(ComponentContext ctxt) {
+
+        try {
+            OAuth2CustomAuthenticator oAuth2CustomAuthenticator = new OAuth2CustomAuthenticator();
+            Hashtable<String, String> props = new Hashtable<>();
+            ctxt.getBundleContext().registerService(ApplicationAuthenticator.class.getName(), oAuth2CustomAuthenticator,
+                    props);
+            if (logger.isDebugEnabled()) {
+                logger.debug("----OAuth2 Custom Authenticator is activated----");
```
Comment (piraveena) [inline authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java:46]:
```suggestion
                logger.debug("OAuth2 Custom Authenticator is activated");
```

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            }
+
+        } catch (Throwable e) {
+            logger.error("----Error while activating OAuth2 Custom Authenticator----", e);
```
Comment (piraveena) [inline authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java:50]:
```suggestion
            logger.error("Error while activating OAuth2 Custom Authenticator", e);
```

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        }
+    }
+
+    @Deactivate
+    protected void deactivate(ComponentContext ctxt) {
+
+        if (logger.isDebugEnabled()) {
+            logger.debug("----OAuth2 Custom Authenticator is deactivated----");
```
Comment (piraveena) [inline authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java:58]:
```suggestion
            logger.debug("OAuth2 Custom Authenticator is deactivated");
```

File: authenticators/pom.xml [hunk 6]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!--
-  ~ Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+  ~ Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (piraveena) [inline authenticators/pom.xml:3]: no need to change the license txt

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        Property clientId = new Property();
+        clientId.setName(OAuth2CustomAuthenticatorConstants.CLIENT_ID);
+        clientId.setDisplayName("Client Id");
+        clientId.setRequired(true);
+        clientId.setDescription("Enter client identifier value");
+        configProperties.add(clientId);
+
+        Property clientSecret = new Property();
+        clientSecret.setName(OAuth2CustomAuthenticatorConstants.CLIENT_SECRET);
+        clientSecret.setDisplayName("Client Secret");
```
Comment (piraveena) [inline authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java:82]: have these as constants in the constant class, same for clientId disply name and call back url

Comment (Avarjana) [reply authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java]: Addressed in https://github.com/wso2/samples-is/pull/338/commits/77e25c91e9c6a5bb5eca3e6b2743eed10d85ace1

File: authenticators/pom.xml (hunk 6 above)
Comment (Avarjana) [reply authenticators/pom.xml]: Addressed in https://github.com/wso2/samples-is/pull/338/commits/77e25c91e9c6a5bb5eca3e6b2743eed10d85ace1

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java (hunk 5 above)
Comment (Avarjana) [reply authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java]: Addressed in https://github.com/wso2/samples-is/pull/338/commits/77e25c91e9c6a5bb5eca3e6b2743eed10d85ace1

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java (hunk 4 above)
Comment (Avarjana) [reply authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java]: Addressed in https://github.com/wso2/samples-is/pull/338/commits/77e25c91e9c6a5bb5eca3e6b2743eed10d85ace1

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java (hunk 3 above)
Comment (Avarjana) [reply authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java]: Addressed in https://github.com/wso2/samples-is/pull/338/commits/77e25c91e9c6a5bb5eca3e6b2743eed10d85ace1

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java (hunk 2 above)
Comment (Avarjana) [reply authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java:52]: Addressed in https://github.com/wso2/samples-is/pull/338/commits/77e25c91e9c6a5bb5eca3e6b2743eed10d85ace1

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/pom.xml (hunk 1 above)
Comment (Avarjana) [reply authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/pom.xml]: Addressed in https://github.com/wso2/samples-is/pull/338/commits/77e25c91e9c6a5bb5eca3e6b2743eed10d85ace1

**Review by piraveena — APPROVED** [review]

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+import org.wso2.carbon.identity.application.authenticator.oauth2.Oauth2GenericAuthenticator;
+import org.wso2.carbon.identity.application.common.model.Property;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.wso2.carbon.identity.sample.oauth2.federated.authenticator.OAuth2CustomAuthenticatorConstants.*;
+
+/***
```
Comment (Yasara-Y) [inline authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java:29]: Formatting issue. Doc comment should look like follows. Add a full stop at the end of each line as well.
/**
 * OAuth2 Custom ...
 */

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+package org.wso2.carbon.identity.sample.oauth2.federated.authenticator;
+
+import org.wso2.carbon.identity.application.authenticator.oauth2.Oauth2GenericAuthenticator;
+import org.wso2.carbon.identity.application.common.model.Property;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.wso2.carbon.identity.sample.oauth2.federated.authenticator.OAuth2CustomAuthenticatorConstants.*;
```
Comment (Yasara-Y) [inline authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java:27]: Avoid * imports. Check other places too.

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        return AUTHENTICATOR_FRIENDLY_NAME;
+    }
+
+    @Override
+    public String getName() {
+
+        return AUTHENTICATOR_NAME;
+    }
+
+    // TODO: Override buildClaims to match the response of respective OAuth2 identity provider
```
Comment (sachiniWettasinghe) [inline authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java:49]: Why do we have a TODO committed? Have we addressed this case later in a PR?

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+        try {
+            OAuth2CustomAuthenticator oAuth2CustomAuthenticator = new OAuth2CustomAuthenticator();
+            Hashtable<String, String> props = new Hashtable<>();
+            ctxt.getBundleContext().registerService(ApplicationAuthenticator.class.getName(), oAuth2CustomAuthenticator,
+                    props);
+            if (logger.isDebugEnabled()) {
+                logger.debug("OAuth2 Custom Authenticator is activated");
+            }
+
```
Comment (sachiniWettasinghe) [inline authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java:48]: Remove redundant line

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    @Activate
+    protected void activate(ComponentContext ctxt) {
+
+        try {
+            OAuth2CustomAuthenticator oAuth2CustomAuthenticator = new OAuth2CustomAuthenticator();
+            Hashtable<String, String> props = new Hashtable<>();
+            ctxt.getBundleContext().registerService(ApplicationAuthenticator.class.getName(), oAuth2CustomAuthenticator,
+                    props);
+            if (logger.isDebugEnabled()) {
+                logger.debug("OAuth2 Custom Authenticator is activated");
```
Comment (sachiniWettasinghe) [inline authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java:46]: Suggestion: "OAuth2 Custom Authenticator is activated."

Comment (sachiniWettasinghe) [reply authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java:46]: Check for similar cases in other places as well.

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java (hunk 10 above)
Comment (Avarjana) [reply authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java:49]: This is to notify the developer to override the buildClaims method and do the implementation as necessary.

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java (hunk 9 above)
Comment (Avarjana) [reply authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java:27]: All the constants are used in this class. Therefore, will not make a change as of now.

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        userInfoUrl.setDisplayName(USER_INFO_URL_DP);
+        userInfoUrl.setRequired(true);
+        userInfoUrl.setDescription(USER_INFO_URL_DESC);
+        userInfoUrl.setDisplayOrder(6);
+        configProperties.add(userInfoUrl);
+
+        return configProperties;
+    }
+}
+
```
Comment (somindatommy) [inline authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java:108]: Remove the extra new line

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticatorConstants.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ ******************************************************************************/
+
+package org.wso2.carbon.identity.sample.oauth2.federated.authenticator;
+
+public class OAuth2CustomAuthenticatorConstants {
```
Comment (somindatommy) [inline authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticatorConstants.java:21]: Missing the class comment

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java (hunk 13 above)
Comment (somindatommy) [reply authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java:108]: Fix in other files as well

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java (hunk 12 above)
Comment (Avarjana) [reply authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java:46]: Fixed with https://github.com/wso2/samples-is/pull/359

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java (hunk 11 above)
Comment (Avarjana) [reply authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java:48]: Fixed with https://github.com/wso2/samples-is/pull/359

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java (hunk 8 above)
Comment (Avarjana) [reply authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java:29]: Fixed with https://github.com/wso2/samples-is/pull/359

File: authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java (hunk 13 above)
Comment (Avarjana) [reply authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/OAuth2CustomAuthenticator.java:108]: Fixed with https://github.com/wso2/samples-is/pull/360

**Pull request merged**